[FIX] Maskwood Nexus not effecting MDFCs

grimreap124 opened the pull request:
Updating Maskwood Nexus to correctly apply to MDFCs.

Tutoring with Magda, Brazen Outlaw wouldn't allow you to get them when Maskwood was in play

xenohedron commented on Mage.Sets/src/mage/cards/m/MaskwoodNexus.java:
```diff
@@ -81,31 +84,42 @@ public boolean apply(Game game, Ability source) {
-                game.getState().getCreateMageObjectAttribute(card, game).getSubtype().setIsAllCreatureTypes(true);
-            }
-        }
+        controller.getHand().stream().map(game::getCard).filter(Objects::nonNull).filter(card -> card.isOwnedBy(controller.getId()) && card.isCreature(game))
```
Lots of duplicated code here. I know the original had duplicated code, but with the changes it's even worse.

Please try to streamline it e.g. by adding cards from all relevant zones to one set and by factoring out the common filter.

grimreap124 replied:
Done

xenohedron reviewed (commented):
Other notes:
1. Are any other cards potentially in scope for the same bug?
2. Is it possible to write a unit test that demonstrates the fix?

grimreap124 commented:
> Other notes:
> 
>     1. Are any other cards potentially in scope for the same bug?
> 
>     2. Is it possible to write a unit test that demonstrates the fix?

1. I fixed the only other card that doesn't handle multifaced cards
2. Added a unit test for Pinnacle Monk being tutored with Magda and Pinnacle Monk on the stack with Jade Orb of Dragonkind

JayDi85 commented on Mage.Sets/src/mage/cards/m/MaskwoodNexus.java:
```diff
@@ -128,6 +119,16 @@ public boolean apply(Game game, Ability source) {
+        // Apply to all effected cards
+        affectedCards.stream().map(card -> game.getObject(card.getId())).filter(Objects::nonNull).forEach(mageObject -> {
+                    game.getState().getCreateMageObjectAttribute(mageObject, game).getSubtype().setIsAllCreatureTypes(true);
+                    CardUtil.getObjectParts(mageObject).stream().filter(Objects::nonNull).forEach(objectId ->{
```
That’s wrong. Must filter each part for creature. There are possible lands, instances and other parts in mdf.

grimreap124 replied:
Moved filters

xenohedron commented on Mage.Tests/src/test/java/org/mage/test/cards/single/khm/MaskwoodNexusTest.java:
```diff
@@ -0,0 +1,64 @@
+        // activate Magda and select Pinnicle Monk
+        activateAbility(1, PhaseStep.POSTCOMBAT_MAIN, playerA, "Sacrifice five Treasures:");
+
+
```
set strict choose mode true

grimreap124 replied:
When setting strict mode with Magda it fails for some reason but when I do the exact same test in game its successful.
Any ideas why?

xenohedron replied:
No idea based on that, though it could help if you showed what the failure is

xenohedron commented on Mage.Tests/src/test/java/org/mage/test/cards/single/khm/MaskwoodNexusTest.java:
```diff
@@ -0,0 +1,64 @@
+public class MaskwoodNexusTest extends CardTestPlayerBase {
+
+
+    private static final String maskwood = "Maskwood Nexus";
```
good practice - include comments with rules text of cards used for test

xenohedron replied:
not only maskwood nexus, but the others

xenohedron commented on Mage.Tests/src/test/java/org/mage/test/cards/single/khm/MaskwoodNexusTest.java:
```diff
@@ -0,0 +1,64 @@
+
+    }
+
+    public void comboMowuAndMaskwoodNexus() {
```
mowu?

xenohedron commented on Mage.Sets/src/mage/cards/m/MaskwoodNexus.java:
```diff
@@ -81,31 +82,21 @@ public boolean apply(Game game, Ability source) {
-            }
-        }
+        Set<Card> affectedCards =
+                controller.getGraveyard().stream().map(game::getCard).filter(Objects::nonNull).filter(card -> card.isOwnedBy(controller.getId())).collect(Collectors.toSet());
```
style, use new line for each method call when processing a stream

Also, only exile zone needs the owner check, graveyard/hand/library will always be owned

grimreap124 replied:
I believe exile is checking the owner?

xenohedron commented on Mage.Sets/src/mage/cards/a/AshesOfTheFallen.java:
```diff
@@ -62,7 +64,12 @@ public boolean apply(Game game, Ability source) {
                 for (UUID cardId : controller.getGraveyard()) {
                     Card card = game.getCard(cardId);
                     if (card != null && card.isCreature(game) && !card.hasSubtype(subType, game)) {
```
this one also needs update to check creature on part, not main card (and pretty sure no reason to check if already has the subtype)

xenohedron commented on Mage.Tests/src/test/java/org/mage/test/cards/single/khm/MaskwoodNexusTest.java:
```diff
@@ -13,6 +13,15 @@
+ * Artifact
+ * Creatures you control are every creature type. The same is true for creature spells you control and creature cards you own that aren't on the battlefield.
+ * {3}, {T}: Create a 2/2 blue Shapeshifter creature token with changeling.
+ * @author TheElk801
```
author tag should be you if you wrote the test

grimreap124 replied:
Ooh I thought it was the person who wrote the card. I'll up date it when I'm back

xenohedron commented on Mage.Sets/src/mage/cards/a/AshesOfTheFallen.java:
```diff
@@ -62,7 +64,14 @@ public boolean apply(Game game, Ability source) {
+                        MageObject mageObject = game.getObject(card.getId());
+                        if (mageObject != null) {
+                            CardUtil.getObjectPartsAsObjects(mageObject).forEach(objectPart ->{
+                                if (objectPart.isCreature(game)) {
```
So you check isCreature here on the object part, is it really correct to check it earlier on the card also?

xenohedron commented on Mage.Sets/src/mage/cards/m/MaskwoodNexus.java:
```diff
@@ -81,31 +82,31 @@ public boolean apply(Game game, Ability source) {
+        Set<Card> affectedCards =
+                controller.getGraveyard().stream()
+                        .map(game::getCard).filter(Objects::nonNull)
+                        .filter(card -> card.isOwnedBy(controller.getId()) && card.isCreature(game))
```
no need for owner check on zones which will only contain owned cards (graveyard, hand, library)

and as mentioned elsewhere, is it really correct to check if the card is a creature, since you're checking later whether the object part is a creature?
